pkg/server/registry/apigroups/acorn/images/copy.go

```diff
@@ -0,0 +1,353 @@
+	progress := make(chan ggcrv1.Update)
+	// progress gets closed by remote.WriteIndex, so this second channel is so that
+	// we can control closing the result channel in case we need to write an error.
+	progress2 := make(chan ImageProgress)
+	destOpts = append(destOpts, remote.WithProgress(progress))
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		for update := range progress {
+			var errString = ""
+			if update.Error != nil {
+				errString = update.Error.Error()
+			}
+			progress2 <- ImageProgress{
+				Total:       update.Total,
+				Complete:    update.Complete,
+				Error:       errString,
+				CurrentTask: fmt.Sprintf("Copying %s to %s", sourceImage, destImage),
+			}
+		}
+	}()
+
+	go func() {
+		defer func() {
+			wg.Wait()
+			close(progress2)
+		}()
+
+		// Don't write error to chan because it already gets sent to the progress chan by remote.WriteIndex()
+		if err := remote.WriteIndex(destRef, sourceIndex, destOpts...); err != nil {
+			handleWriteIndexError(err, progress)
+		}
+	}()
+
+	return typed.Every(500*time.Millisecond, progress2), nil
+}
+
+func (i *ImageCopy) RepoCopy(ctx context.Context, namespace, source, dest string, sourceAuth, destAuth *apiv1.RegistryAuth, force bool) (<-chan ImageProgress, error) {
+	sourceRepo, err := name.NewRepository(source)
+	if err != nil {
+		return nil, err
+	}
+
+	destRepo, err := name.NewRepository(dest)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceOpts, err := images.GetAuthenticationRemoteOptionsWithLocalAuth(ctx, sourceRepo.Registry, sourceAuth, i.client, namespace, i.transportOpt)
+	if err != nil {
+		return nil, err
+	}
+
+	destOpts, err := images.GetAuthenticationRemoteOptionsWithLocalAuth(ctx, destRepo.Registry, destAuth, i.client, namespace, i.transportOpt)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceTags, err := remote.List(sourceRepo, sourceOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceIndexes := make([]ggcrv1.ImageIndex, len(sourceTags))
+	for i, tag := range sourceTags {
+		sourceIndexes[i], err = remote.Index(sourceRepo.Tag(tag), sourceOpts...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Don't copy tags that already exist in the destination if force is not set
+	if !force {
+		var newIndexSlice []ggcrv1.ImageIndex
+		for i, imageIndex := range sourceIndexes {
+			if err := errIfImageExistsAndIsDifferent(imageIndex, destRepo.Tag(sourceTags[i]), destOpts); err == nil {
+				newIndexSlice = append(newIndexSlice, imageIndex)
+			}
+		}
+		sourceIndexes = newIndexSlice
+	}
+
+	// metachannel is used to send another channel with updates for each tag to be copied
+	metachannel := make(chan chan ggcrv1.Update)
+	// progress is the channel returned by this function and used to write websocket messages to the client
+	progress := make(chan ImageProgress)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	// We create an RWMutex for the currentTag variable, since it is used by both goroutines
+	m := sync.RWMutex{}
+	currentTag := ""
```

Comment:
This, along with the two goroutines that follow, is definitely the grossest part of this code. Would it be worth it to split it out into properly defined functions with parameters, instead of just closures like they are now?

I hate that I had to create a mutex for this, but a shared variable was the only way I could come up with for one goroutine to communicate to the other which tag was currently being copied.